packages/runtime/runtime-utils/src/summarizerNodeUtils.ts

```diff
@@ -0,0 +1,253 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert } from "@fluidframework/common-utils";
+import {
+    ISnapshotTree,
+    IDocumentAttributes,
+    ISequencedDocumentMessage,
+    SummaryType,
+} from "@fluidframework/protocol-definitions";
+import { ISummaryTreeWithStats } from "@fluidframework/runtime-definitions";
+import { SummaryTreeBuilder } from "./summaryUtils";
+
+const baseSummaryTreeKey = "_baseSummary";
+const outstandingOpsBlobKey = "_outstandingOps";
+const maxDecodeDepth = 100;
+
+/** Reads a blob from storage and parses it from JSON. */
+export type ReadAndParseBlob = <T>(id: string) => Promise<T>;
+
+/**
+ * Fetches the sequence number of the snapshot tree by examining the protocol.
+ * @param tree - snapshot tree to examine
+ * @param readAndParseBlob - function to read blob contents from storage
+ * and parse the result from JSON.
+ */
+export async function seqFromTree(
+    tree: ISnapshotTree,
+    readAndParseBlob: ReadAndParseBlob,
+): Promise<number> {
+    const attributesHash = tree.trees[".protocol"].blobs.attributes;
+    const attrib = await readAndParseBlob<IDocumentAttributes>(attributesHash);
+    return attrib.sequenceNumber;
+}
+
+/** Path for nodes in a tree with escaped special characters */
+export class EscapedPath {
+    private constructor(public readonly path: string) { }
+    public static create(path: string): EscapedPath {
+        return new EscapedPath(encodeURIComponent(path));
+    }
+    public static createAndConcat(pathParts: string[]): EscapedPath {
+        let ret = EscapedPath.create(pathParts[0] ?? "");
+        for (let i = 1; i < pathParts.length; i++) {
+            ret = ret.concat(EscapedPath.create(pathParts[i]));
+        }
+        return ret;
+    }
+    public toString(): string {
+        return this.path;
+    }
+    public concat(path: EscapedPath): EscapedPath {
+        return new EscapedPath(`${this.path}/${path.path}`);
+    }
+}
+
+/** Information about a summary relevant to a specific node in the tree */
+export class SummaryNode {
```

Comment:
Comments on each member would be awesome

Reply:
Let me know if you think they could use more detail.